adding AK8 variables

vivannguyen opened the pull request:
Adding the AK8 jet selection and AK8 jet related variables

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -746,23 +793,26 @@ def MuonsFromLQ(T):
 	for n in range(T.nMuon):
 		m = TLorentzVector()
-                m.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+                m.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -746,23 +793,26 @@ def MuonsFromLQ(T):
+			if motherid not in [42,-42]:
+				continue
+			m = TLorentzVector()
+			m.SetPtEtaPhiM(T.GenPart_pt[n],T.GenPart_eta[n],T.GenPart_phi[n],T.GenPart_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -800,21 +850,24 @@ def JetsFromLQ(T):
 		if  ((T.Jet_jetId[n] & 0x2) > 0) and T.Jet_pt[n]>30 and abs(T.Jet_eta[n])<2.4 : #morse only use jets that pass tightid, fixme check jet eta
 			m = TLorentzVector()
-			m.SetPtEtaPhiM(T.Jet_pt[n],T.Jet_eta[n],T.Jet_phi[n],0)
+			m.SetPtEtaPhiM(T.Jet_pt[n],T.Jet_eta[n],T.Jet_phi[n],T.Jet_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -800,21 +850,24 @@ def JetsFromLQ(T):
+			if motherid not in [42,-42]:
+				continue
+			m = TLorentzVector()
+			m.SetPtEtaPhiM(T.GenPart_pt[n],T.GenPart_eta[n],T.GenPart_phi[n],T.GenPart_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -846,45 +899,72 @@ def getLeptonEventFlavorGEN(T):
-			#if (isEleEvent or isMuEvent or isTauEvent) : gotIt=True
+
+	if isData==False:
+
```
in general, let's try to minimize blank lines

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -846,45 +899,72 @@ def getLeptonEventFlavorGEN(T):
 	#         Return TLorentzVectors of the gen and reco muons, and the indices for
 	#         the recomuons as well.
+
+	if radius == 4:
```
While it shouldn't happen, let's protect ourselves against human error. Currently only choices are radius=4 or 8. What if someone in the future puts 6? You can make e.g. a last 'elif' which prints an informative error and kills the whole job. As a general point you should always think about all the possible error modes of a function - case in point these acosine functions that were throwing errors when the input was out of range....

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -893,37 +973,37 @@ def LeptonsAndJetsFromHH(T):
 	for n in range(len(T.Muon_pt)):
 		m = TLorentzVector()
-		m.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+		m.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -893,37 +973,37 @@ def LeptonsAndJetsFromHH(T):
+
+			# AH: here we keep also gmuon status 23 and status 1, of which mother 25. But we do not keep muon status 1 of which mother is 13, -13
+			m = TLorentzVector()
+			m.SetPtEtaPhiM(T.GenPart_pt[n],T.GenPart_eta[n],T.GenPart_phi[n],T.GenPart_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -940,41 +1020,33 @@ def LeptonsAndJetsFromHH(T):
 	for n in range(T.nElectron):
 		m = TLorentzVector()
-		m.SetPtEtaPhiM(T.Electron_pt[n],T.Electron_eta[n],T.Electron_phi[n],0)
+		m.SetPtEtaPhiM(T.Electron_pt[n],T.Electron_eta[n],T.Electron_phi[n],T.Electron_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -940,41 +1020,33 @@ def LeptonsAndJetsFromHH(T):
+				continue
+			if motherid in [23,-23]: onShellZEle=True
+			m = TLorentzVector()
+			m.SetPtEtaPhiM(T.GenPart_pt[n],T.GenPart_eta[n],T.GenPart_phi[n],T.GenPart_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1006,50 +1077,43 @@ def LeptonsAndJetsFromHH(T):
+			if motherid not in [23,-23,25,-25] :#parentParticles:
+				continue
+			m = TLorentzVector()
+			m.SetPtEtaPhiM(T.GenPart_pt[n],T.GenPart_eta[n],T.GenPart_phi[n],T.GenPart_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+					sub1 = emptyvector
+				else:
+					sub1 = TLorentzVector()
+					sub1.SetPtEtaPhiM(SubJetPt[SubJetInd1[j]],SubJetEta[SubJetInd1[j]],SubJetPhi[SubJetInd1[j]],SubJetMass[SubJetInd1[j]])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+					sub2 = emptyvector
+				else:
+					sub2 = TLorentzVector()
+					sub2.SetPtEtaPhiM(SubJetPt[SubJetInd2[j]],SubJetEta[SubJetInd2[j]],SubJetPhi[SubJetInd2[j]],SubJetMass[SubJetInd2[j]])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+					sub1 = emptyvector
+				else:
+					sub1 = TLorentzVector()
+					sub1.SetPtEtaPhiM(SubJetPt[SubJetInd1[j]],SubJetEta[SubJetInd1[j]],SubJetPhi[SubJetInd1[j]],SubJetMass[SubJetInd1[j]])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+					sub2 = emptyvector
+				else:
+					sub2 = TLorentzVector()
+					sub2.SetPtEtaPhiM(SubJetPt[SubJetInd2[j]],SubJetEta[SubJetInd2[j]],SubJetPhi[SubJetInd2[j]],SubJetMass[SubJetInd2[j]])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+		gendRujH=highestptGenJetH.DeltaR(highestptGenMuon)
+		gendRujZ=highestptGenJetZ.DeltaR(highestptGenMuon)
+
+	if radius == 4: return([genmuons,matchedrecomuons,recomuoninds,genelectrons,matchedrecoelectrons,recoelectroninds,genjetsH,matchedrecojetsH,recojetindsH,genjetsZ,matchedrecojetsZ,recojetindsZ,onShellZMu,onShellZEle])
```
This looks like a problem in the indentation, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1094,132 +1158,281 @@ def LeptonsAndJetsFromHH(T):
+		gendRujZ=highestptGenJetZ.DeltaR(highestptGenMuon)
+
+	if radius == 4: return([genmuons,matchedrecomuons,recomuoninds,genelectrons,matchedrecoelectrons,recoelectroninds,genjetsH,matchedrecojetsH,recojetindsH,genjetsZ,matchedrecojetsZ,recojetindsZ,onShellZMu,onShellZEle])
+	elif radius == 8: return([jetsH,matchedrecojetH,recojetindsH,matchedrecojetZ,recojetindsZ,dRujH,dRujZ,gendRujH,gendRujZ,CSV_jetH,CSV_jetZ])
```
This looks like a problem in the indentation, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1231,7 +1444,7 @@ def PropagatePTChangeToMET(met,original_object,varied_object):
 
 def LooseIDMuons(T,_met,variation,isdata):
-	# Purpose: Gets the collection of muons passing loose muon ID.
+	# Purpose: Gets the collection of muons passing loose muon ID.f
```
delete the 'f' at the end :)

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1290,24 +1503,30 @@ def LooseIDMuons(T,_met,variation,isdata):
 			OldMu = TLorentzVector()
-			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],0)
-			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1290,24 +1503,30 @@ def LooseIDMuons(T,_met,variation,isdata):
-			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
+			#NewMu.SetPtEtaPhiM(_MuonPt[n],T.MuonEta[n],T.MuonPhi[n],0)
+			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1382,20 +1601,20 @@ def MediumIDMuons(T,_met,variation,isdata):
 		# Don't apply isolation for QCD studies
 		if nonisoswitch != True:
-			Pass *= correctedIso < 0.25
+			Pass *= correctedIso/_MuonPt[n] < 0.25
```
I think the Muon_pfRelIso04_all is already the relative isolation, and therefore does not need to be divided by _MuonPt

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1405,8 +1624,8 @@ def MediumIDMuons(T,_met,variation,isdata):
 			OldMu = TLorentzVector()
-			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],0)
-			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1405,8 +1624,8 @@ def MediumIDMuons(T,_met,variation,isdata):
-			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],0)
-			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],0)
+			NewMu.SetPtEtaPhiM(_MuonPt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
+			OldMu.SetPtEtaPhiM(T.Muon_pt[n],T.Muon_eta[n],T.Muon_phi[n],T.Muon_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1463,7 +1682,7 @@ def mvaWP90Electrons(T,_met,variation,isdata):
 		Pass *= (barrel+endcap)
 
-	        Pass *= T.Electron_mvaFall17V2noIso_WP90[n]>0
+	        Pass *= T.Electron_mvaFall17V2Iso_WP90[n]>0
```
We should understand if this is supposed to be noIso or Iso. I thought noIso, since we apply the isolation separately later. This is important when we want to do the QCD estimate, where we do not apply the isolation.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1472,22 +1691,55 @@ def mvaWP90Electrons(T,_met,variation,isdata):
 			if (barrel):
 				#Pass *= iso<0.0588#tight
-				Pass *= T.Electron_mvaFall17V2Iso_WP90[n]>0
+				Pass *= T.Electron_mvaFall17V2noIso_WP90[n]>0
```
same comment about noIso

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1472,22 +1691,55 @@ def mvaWP90Electrons(T,_met,variation,isdata):
 			elif (endcap):
 				#Pass *= iso<0.0571
-				Pass *= T.Electron_mvaFall17V2Iso_WP90[n]>0
+				Pass *= T.Electron_mvaFall17V2noIso_WP90[n]>0
```
same comment about noIso

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -1914,7 +2166,7 @@ def JERModifiedPt(pt,eta,phi,T,modtype):
 	for n in range(len(T.GenJet_pt)):
 		gjet = TLorentzVector()
-		gjet.SetPtEtaPhiM(T.GenJet_pt[n],T.GenJet_eta[n],T.GenJet_phi[n],0.0)
+		gjet.SetPtEtaPhiM(T.GenJet_pt[n],T.GenJet_eta[n],T.GenJet_phi[n],T.GenJet_mass[n])
```
This looks like a problem in the indentation compared to the old version, please check.

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -3591,29 +4262,52 @@ def FullKinematicCalculation(T,variation):
+	toreturn += [_ak8_ptHj,_ak8_ptZj,_ak8_massHj,_ak8_massZj]
+	toreturn += [_ak8_ptHj_Matched,_ak8_ptZj_Matched,_ak8_massHj_Matched,_ak8_massZj_Matched]
+
+	toreturn += [_ak8_Muujj,_ak8_Meejj]
```
don't think we need Muujj for HH

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -185,30 +186,50 @@
+_kinematicvariables += ['AK8_Pt_Hjet','AK8_Pt_Zjet','AK8_mass_Hjet','AK8_mass_Zjet']
+_kinematicvariables += ['AK8_Pt_Hjet_Matched','AK8_Pt_Zjet_Matched','AK8_mass_Hjet_Matched','AK8_mass_Zjet_Matched']
+
+_kinematicvariables += ['AK8_M_uujj','AK8_M_eejj']
```
don't think we need Muujj for HH

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -3785,8 +4478,8 @@ def MetVector(T):
 # Please don't edit here. It is static. The kinematic calulations are the only thing to edit!
-#lumisection = array.array("L",[0])
-#t.SetBranchAddress("ls",lumisection)
+lumisection = array.array("L",[0])
```
this should be commented out

DavidMorse commented on NTupleAnalyzerHH_nanoAOD.py:
```diff
@@ -3785,8 +4478,8 @@ def MetVector(T):
-#lumisection = array.array("L",[0])
-#t.SetBranchAddress("ls",lumisection)
+lumisection = array.array("L",[0])
+t.SetBranchAddress("ls",lumisection)
```
these two lines should be commented out

vivannguyen commented:
Removed some blank lines; added error statement and exit for invalid radius arguments in LeptonsAndJetsFromHH; removed division by _MuonPt for correctedIso; changed Electron_mvaFall17V2Iso_WP90 to nonIso; removed Muujj and Meejj variables for AK8 jets
